vhost_user_blk: implement support for direct (O_DIRECT) mode

slp opened the pull request:
Add support for opening the disk images with O_DIRECT. This allows
bypassing the host's file system cache, which is useful to avoid
polluting its cache and for better data integrity.

On the downside, O_DIRECT imposes certain alignment restrictions for
the I/O operations done on the disk image. Ideally, a layer between
the disk format and the underlying file should be implemented to cope
with such restrictions. But, for the moment, we just reject opening
qcow2 images in direct mode.

slp commented:
> We plan to make CH binary be both the vhost_user_blk/net backend as well as the VMM itself so we can self activate. EIther by forking or maybe just spawning /proc/<pid>/exe again so it would be better if we could keep the parameters contained together.

Oh, I see. I saw that vhost-user-fs now uses independent arguments and assumed it would be okay for vhost-user-blk too. I can add "direct=bool" as an option for the "backend" argument, if that's better.
 
> Is there any reason why we wouldn't want to always use `O_DIRECT` for the raw file access? Having the blocks cached in the host doesn't seem useful when the guest will also do caching anyway.
> 
> As such i'd propose a simpler PR that just sets `O_DIRECT` always for raw files.

There are some filesystems that don't support O_DIRECT, and sometimes it's nice being able to abuse the host's cache to have a significant (though risky) boost in I/O performance. I think it'd be better to have both options for raw devices.

rbradford commented:
Thanks @slp. A `direct=` option would be best. Could we default to `O_DIRECT` to being on?

slp commented:
> Thanks @slp. A `direct=` option would be best. Could we default to `O_DIRECT` to being on?

Sure, but I think it'd be better to do that after it's supported for qcow2 images too. Otherwise it'd probably be confusing for users ("qcow2 suddenly became way faster than raw?").

rbradford commented:
@slp okay, fair enough.

yangzhon commented:
@slp @rbradford 

We can enable WCE feature in the backend to avoid add "direct=bool" in the backend. In fact, the current guest virtio_blk.c driver only support "write back" and "write through" cache policy and "none" cache mode should be supported in this driver.

for "O_DIRECT" in the backend, which can be changed by " echo none > /sys/block/vda/cache_type" in the guest. But from current qemu backend, which is changed by "echo write through > /sys/block/vda/cache_type". If this is one bug, i will send patch to qemu and linux kernel community after current work is over. thanks.

slp commented:
@yangzhon 

> We can enable WCE feature in the backend to avoid add "direct=bool" in the backend. In fact, the current guest virtio_blk.c driver only support "write back" and "write through" cache policy and "none" cache mode should be supported in this driver.
> 

That would mean that the guest would decide whether the host's filesystem cache should be used or not, which I don't think it's what most users would want.

As a reference, in QEMU the only effect that WCE has is (if disabled) making the block layer issue some extra flushes to the underlying file or block device.

slp commented:
> Add an integration test to turn direct on?

Sure, how should it look like? Maybe just like `test_boot_from_vhost_user_blk` (which already uses a raw disk) but with "direct=true"? Something like `test_boot_from_vhost_user_blk_direct`?

rbradford commented:
> > Add an integration test to turn direct on?
> 
> Sure, how should it look like? Maybe just like `test_boot_from_vhost_user_blk` (which already uses a raw disk) but with "direct=true"? Something like `test_boot_from_vhost_user_blk_direct`?

Adding another test that just varies it direct would be best. I like your choice of name.

yangzhon commented:
@slp 
I used the fio tool only to check performance, the command as below
(1).read
sudo fio --ioengine=libaio --filename=/dev/vdb --direct=1 --rw=randread --bs=4k --size=1G --numjobs=64 --runtime=10 --group_reporting --name=file2
(2). write
sudo fio --ioengine=libaio --filename=/dev/vdb --direct=1 --rw=randwrite --bs=4k --size=1G --numjobs=64 --runtime=10 --group_reporting --name=file2

Test result
(1). rust device + rust backend
read: iops=173
write: iops=460

(2) qemu device + qemu backend
read: iops=190
write: iops=458

(3). qemu + spdk(but cat /sys/block/vda/cache_type is "write back")
read: iops=203313
write: iops=199622

yangzhon commented:
@slp ,

Please help check if your patch omitted below cmdline setting? thanks
--- a/src/bin/vhost_user_blk.rs
+++ b/src/bin/vhost_user_blk.rs
@@ -290,7 +290,7 @@ fn main() {
                 .long("backend")
                 .help(
                     "Backend parameters \"image=<image_path>,\
-                     sock=<socket_path>\"",
+                     sock=<socket_path>, direct=true|false\"",
                 )

sboeuf commented:
@slp any progress here? Let us know if you run into any trouble writing the test :)

slp commented:
@sboeuf Yeah, while running the tests, I've noticed some issues when opening the image with O_DIRECT. Turns out that, while Linux seems to always provide a sector-aligned address in the data descriptor (probably a side-effect of using page-aligned buffers for scatter-gather lists in case there's DMA involved), hypervisor-fw doesn't do that.

I've checked the VirtIO spec, and it doesn't put any restrictions to data descriptor addresses, so what hypervisor-fw does is perfectly fine (thought probably inefficient). So we need to come up with a layer that would sit between the format image (RawFile/QcowFile) and the File. That's definitely doable, but will make this patch series larger and more complex. On the bright side, we'll get O_DIRECT for qcow2 files too. :-D

My intention is to update this PR sometime between this week and the next one.

sboeuf commented:
@slp thanks for the update!

>while running the tests, I've noticed some issues when opening the image with O_DIRECT. Turns out that, while Linux seems to always provide a sector-aligned address in the data descriptor (probably a side-effect of using page-aligned buffers for scatter-gather lists in case there's DMA involved), hypervisor-fw doesn't do that. I've checked the VirtIO spec, and it doesn't put any restrictions to data descriptor addresses, so what hypervisor-fw does is perfectly fine (thought probably inefficient). So we need to come up with a layer that would sit between the format image (RawFile/QcowFile) and the File.

/cc @rbradford

rbradford commented:
The problem is that Rust doesn't offer a way (AFAICT) to allow us to allocate our buffers on the stack with a particular alignment.

slp commented:
> The problem is that Rust doesn't offer a way (AFAICT) to allow us to allocate our buffers on the stack with a particular alignment. (The buffers that the firmware uses with virtio block are stack allocated.)

Yeah, Rust seems pretty limited in that aspect. Thought I think I read somewhere they're working on improving that, given the interest Rust has sparked around people doing bare metal stuff.

Anyway, I'm not blaming hypervisor-fw here. As I said, the VirtIO spec doesn't say anything about alignment requests for those addresses, so this is something that must be fixed in vhost-user-blk.

slp commented:
Sorry for the long wait. I've updated the patchset to include changes in RawFile to be able to deal with alignment restrictions. I also changed QcowFile to use RawFile instead of File, so we can abstract low-level peculiarities, like this one. PTAL when you have a minute.

Do we still want to make O_DIRECT the default behaviour, now that it's also supported for qcow2 images? Do we also want to add O_DIRECT support to device_manager.rs?

sboeuf commented:
@slp 
>Sorry for the long wait. I've updated the patchset to include changes in RawFile to be able to deal with alignment restrictions. I also changed QcowFile to use RawFile instead of File, so we can abstract low-level peculiarities, like this one. PTAL when you have a minute.

Great news, so just to clarify, you don't run into the issues that you were seeing with the `hypervisor-fw` anymore, right?

>Do we still want to make O_DIRECT the default behaviour, now that it's also supported for qcow2 images? Do we also want to add O_DIRECT support to device_manager.rs?

For now, let's not make this option the default as it might not work with every file.
Could you also run some performance test to assess what is the performance gain with this feature enabled?

We're gonna review your PR, and I think we'll merge it if it looks fine, as we can have a follow up PR for making this the default.

slp commented:
@sboeuf 

> > Sorry for the long wait. I've updated the patchset to include changes in RawFile to be able to deal with alignment restrictions. I also changed QcowFile to use RawFile instead of File, so we can abstract low-level peculiarities, like this one. PTAL when you have a minute.
> 
> Great news, so just to clarify, you don't run into the issues that you were seeing with the `hypervisor-fw` anymore, right?

Nope, CH is now able to boot properly jumping through hypervisor-hw, and the corresponding CI tests pass properly.

> > Do we still want to make O_DIRECT the default behaviour, now that it's also supported for qcow2 images? Do we also want to add O_DIRECT support to device_manager.rs?
> 
> For now, let's not make this option the default as it might not work with every file.

OK.

> Could you also run some performance test to assess what is the performance gain with this feature enabled?

There isn't going to be a performance gain from this. There's an important (in fact, critical for production usage) advantage though, as with O_DIRECT you forbid the guests from polluting the host's filesystem cache. On one hand, you save memory, as otherwise data would be cached twice (once in the guest's fs cache, and other in the host's fs cache), and on the other, you also stop a malicious or misbehaving guest from having in potential impact on the server.

It's also useful for doing realistic performance benchmarks from the guest, as otherwise you're just relying on the host's cache, and everything depends on your luck hitting an already cached block.

> We're gonna review your PR, and I think we'll merge it if it looks fine, as we can have a follow up PR for making this the default.

Nice. I think it'd be also nice having this option for normal (attached via device_manager) disks, but, as you said, we can add that later.

slp commented:
CI fails because I forgot to update the tests in qcow.rs. Will do that later today.

rbradford commented on vm-virtio/src/block.rs:
```diff
@@ -94,42 +96,311 @@ impl ExecuteError {
+                    to_copy.try_into().unwrap(),
+                );
+                dealloc(ptr, layout)
+            };
```
Could we use `slice::from_raw_parts()` and `buf.copy_from_slice()` instead of memcpy?

slp replied:
This is weird, I'd swear I tried this approach when I started working on this (before Xmas) and hit some issue, but I'm now reviewing the code and can't figure out what the problem could have been.

I'm going to give it another try. Let's see how it works.

slp replied:
I've updated the patch series and it works nicely, not sure what kind of issue I hit last time. I also switched the aligned cases to use `self.file.[read|write]`, saving a couple `unsafe` blocks.

I'm still using `pread/pwrite`, as it saves a `lseek` call, but once `read_at/write_at` is stabilized we can cut down another three `unsafe` blocks.

rbradford commented on vm-virtio/src/block.rs:
```diff
@@ -94,42 +96,311 @@ impl ExecuteError {
-        self.file.read(buf)
-    }
-}
+        if self.is_aligned(buf) {
```
Maybe add a comment

`// Without O_DIRECT buffers are always aligned`

slp replied:
Sure, makes sense.

slp commented:
@rbradford 

> Do you think it would be viable to also do this for virtio-block?

Sure, in fact I have a patch for that, but I dropped it from the series since I wasn't sure that want it there. I can easily rebase it on top of the current series.

rbradford commented:
> @rbradford
> 
> > Do you think it would be viable to also do this for virtio-block?
> 
> Sure, in fact I have a patch for that, but I dropped it from the series since I wasn't sure that want it there. I can easily rebase it on top of the current series.

@slp cool, we can put it in a new PR after this one is merged.

My approval still stands.

rbradford commented:
@slp CI failed due to a clippy issue.

rbradford commented:
> @slp CI failed due to a clippy issue.

I've pushed a fix as i'd like to see this change merged soon
